auth/common.cc

```diff
@@ -113,4 +118,118 @@ ::service::query_state& internal_distributed_query_state() noexcept {
     return qs;
 }
 
+future<::service::group0_guard> start_group0_operation(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        seastar::abort_source* as) {
+    co_return co_await qp.container().invoke_on(0,
+        [&group0_client, as](cql3::query_processor& qp) -> future<::service::group0_guard> {
+            co_return co_await group0_client.start_operation(as);
+    });
+}
+
+static future<> announce_mutations_with_guard(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        std::vector<canonical_mutation> muts,
+        ::service::group0_guard group0_guard,
+        seastar::abort_source* as) {
+    co_await qp.container().invoke_on(0,
+        [&group0_client, muts = std::move(muts), group0_guard = std::move(group0_guard), as]
+        (cql3::query_processor&) mutable -> future<> {
+            auto group0_cmd = group0_client.prepare_command(
+                ::service::write_mutations{
+                    .mutations{std::move(muts)},
+                },
+                group0_guard,
+                "auth: modify internal data"
+            );
+            co_await group0_client.add_entry(std::move(group0_cmd), std::move(group0_guard), as);
+    });
+}
+
+static future<> announce_mutations_maybe_split(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        std::vector<canonical_mutation> muts,
+        ::service::group0_guard group0_guard,
+        seastar::abort_source* as) {
+    auto begin = muts.begin();
+    auto end = muts.end();
+    size_t batch_size  = muts.size();
+    auto remaining = batch_size;
+    while (remaining) {
+        try {
+            co_await announce_mutations_with_guard(qp, group0_client, {begin, end}, std::move(group0_guard), as);
+            remaining -= (end - begin);
+            begin = end;
+        } catch (raft::command_is_too_big_error&) {
+            batch_size /= 2;
+            if (batch_size == 0) {
+                std::throw_with_nested(std::runtime_error("raft add entry failed with batch size 1"));
+            }
+            // when retrying we don't change begin and end gets truncated below
+        }
+        if (remaining) {
+            end = begin + std::min(batch_size, remaining);
+            // FIXME: we need a way to reuse single guard when retry
+            // doesn't require command data regeneration as in this case
+            // otherwise we may violate operation consistency guarantee
+            group0_guard = co_await start_group0_operation(qp, group0_client, as);
```

Comment:
> ok, so new_group0_state_id will be the ID **after** entry is committed?

Yes.

> But what can I do with partial data which is already committed if I discover state change in between?

This probably depends on a particular operation. For example, if you are revoking permissions, then it would make sense to me to recalculate the permissions that need to be revoked and retry (not all work is lost, you won't have to revoke the permissions you committed in the previous attempt).

What additional guarantees are you aiming to provide here? Make sure that each operation is executed in full, even if it requires multiple steps?

Reply:
> permissions you committed in the previous attempt

so it can be that I am deleting role permissions because someone deleted related role, but meanwhile someone added a new role with the same name, then final result will be bad anyway.

> What additional guarantees are you aiming to provide here 

With the FIXME I wanted to convey that ideally:
- multi command operation starts, takes a lock or something
- before it finishes all new operations receive concurrent modification exception
- multi command operation finishes successfully
- new operations can be retried either by user or internal layer

bonus: ideally the lock is on auth tables only not the whole group0

But I think that can be left for the future because it's super edge case - involves having hundreds permissions/roles and modifying auth concurrently - which rarely makes sense outside of testing env.

Current code guarantee is that:
- in a typical case writes are atomic, under guard
- when it needs multiple commands it stops being atomic - it's anyway not worse than current implementation